File: curseforge.py
import time
import requests
from typing import Optional
from models import ModVersion

BASE_URL = "https://api.curseforge.com/v1"

_LOADER_TYPE = {
    "forge": 1,
    "fabric": 4,
    "quilt": 5,
    "neoforge": 6,
}

_LOADER_NAME = {v: k for k, v in _LOADER_TYPE.items()}


def _headers(api_key: str) -> dict:
    return {
        "x-api-key": api_key,
        "Accept": "application/json",
    }


def _request_with_retry(method: str, url: str, api_key: str, **kwargs) -> requests.Response:
    resp = requests.request(method, url, headers=_headers(api_key), **kwargs)
    if resp.status_code == 429:
        time.sleep(int(resp.headers.get("Retry-After", 60)))
        resp = requests.request(method, url, headers=_headers(api_key), **kwargs)
    resp.raise_for_status()
    return resp


def parse_file_object(raw: dict) -> ModVersion:
    loader_type = raw.get("modLoaderType", 0)
    loader_name = _LOADER_NAME.get(loader_type, str(loader_type))
    return ModVersion(
        version_id=str(raw["id"]),
        version_number=raw.get("displayName", str(raw["id"])),
        filename=raw.get("fileName", ""),
        download_url=raw.get("downloadUrl") or "",
        game_versions=raw.get("gameVersions", []),
        loaders=[loader_name],
        source="curseforge",
    )


def get_versions_by_fingerprint(fingerprints: list, api_key: str) -> dict:
    """Return dict mapping fingerprint (int) → file dict for exact matches."""
    if not fingerprints:
        return {}
    resp = _request_with_retry(
        "POST",
        f"{BASE_URL}/fingerprints",
        api_key,
        json={"fingerprints": fingerprints},
    )
    data = resp.json().get("data", {})
    result = {}
    for match in data.get("exactMatches", []):
        fp = match.get("id")
        file_data = match.get("file")
        if fp is not None and file_data:
            result[fp] = file_data
    return result


def get_mod_info(mod_id: int, api_key: str) -> dict:
    try:
        resp = _request_with_retry("GET", f"{BASE_URL}/mods/{mod_id}", api_key)
        return resp.json().get("data", {})
    except requests.HTTPError:
        return {}


def get_mods_info(mod_ids: list, api_key: str) -> dict:
    """Batch fetch mod metadata. Returns dict mod_id (int) → mod dict."""
    if not mod_ids:
        return {}
    try:
        resp = _request_with_retry(
            "POST",
            f"{BASE_URL}/mods",
            api_key,
            json={"modIds": mod_ids},
        )
        return {m["id"]: m for m in resp.json().get("data", [])}
    except requests.HTTPError:
        return {}


def get_latest_version(mod_id: int, minecraft_version: str, loader: str, api_key: str) -> Optional[ModVersion]:
    loader_type = _LOADER_TYPE.get(loader.lower(), 0)
    try:
        resp = _request_with_retry(
            "GET",
            f"{BASE_URL}/mods/{mod_id}/files",
            api_key,
            params={
                "gameVersion": minecraft_version,
                "modLoaderType": loader_type,
                "pageSize": 1,
                "sortOrder": "desc",
            },
        )
        files = resp.json().get("data", [])
        if not files:
            return None
        return parse_file_object(files[0])
    except requests.HTTPError:
        return None

File: rst/src/curseforge.rs
use anyhow::Result;
use reqwest::Client;
use serde::Deserialize;
use serde_json::json;
use std::collections::HashMap;
use crate::models::{ModSource, ModVersion};

const BASE_URL: &str = "https://api.curseforge.com/v1";

fn loader_id(loader: &str) -> u32 {
    match loader {
        "forge" => 1,
        "fabric" => 4,
        "quilt" => 5,
        "neoforge" => 6,
        _ => 0,
    }
}

#[derive(Deserialize)]
struct CfFile {
    id: u32,
    #[serde(rename = "displayName")]
    display_name: String,
    #[serde(rename = "fileName")]
    file_name: String,
    #[serde(rename = "downloadUrl")]
    download_url: Option<String>,
    #[serde(rename = "gameVersions")]
    game_versions: Vec<String>,
    #[serde(rename = "fileFingerprint", default)]
    file_fingerprint: u64,
}

#[derive(Deserialize)]
struct CfMod {
    id: u32,
    name: String,
}

#[derive(Deserialize)]
struct CfFingerprintMatch {
    id: u32,
    file: CfFile,
}

#[derive(Deserialize)]
struct CfFingerprintData {
    #[serde(rename = "exactMatches")]
    exact_matches: Vec<CfFingerprintMatch>,
}

#[derive(Deserialize)]
struct CfResponse<T> {
    data: T,
}

async fn request_with_retry(client: &Client, api_key: &str, url: &str, body: Option<serde_json::Value>) -> Result<reqwest::Response> {
    loop {
        let req = if let Some(ref b) = body {
            client.post(url).header("x-api-key", api_key).json(b)
        } else {
            client.get(url).header("x-api-key", api_key)
        };
        let resp = req.send().await?;
        if resp.status().as_u16() == 429 {
            let retry_after = resp
                .headers()
                .get("Retry-After")
                .and_then(|v| v.to_str().ok())
                .and_then(|v| v.parse::<u64>().ok())
                .unwrap_or(1);
            tokio::time::sleep(std::time::Duration::from_secs(retry_after)).await;
            continue;
        }
        return Ok(resp);
    }
}

fn parse_cf_file(file: CfFile, _mod_id: u32, _loader: &str) -> Option<ModVersion> {
    let url = file.download_url?;
    let loaders: Vec<String> = file
        .game_versions
        .iter()
        .filter(|v| !v.chars().next().map(|c| c.is_ascii_digit()).unwrap_or(false))
        .map(|v| v.to_lowercase())
        .collect();
    let game_versions: Vec<String> = file
        .game_versions
        .iter()
        .filter(|v| v.chars().next().map(|c| c.is_ascii_digit()).unwrap_or(false))
        .cloned()
        .collect();

    Some(ModVersion {
        version_id: file.id.to_string(),
        version_number: file.display_name,
        filename: file.file_name,
        download_url: url,
        game_versions,
        loaders,
        source: ModSource::CurseForge,
    })
}

/// Returns map from murmur2_fingerprint -> (project_id as string, ModVersion).
pub async fn get_versions_by_fingerprint(
    client: &Client,
    api_key: &str,
    fingerprints: &[u32],
    loader: &str,
) -> Result<HashMap<u32, (String, ModVersion)>> {
    if fingerprints.is_empty() {
        return Ok(HashMap::new());
    }
    let url = format!("{BASE_URL}/fingerprints/432");
    let body = json!({ "fingerprints": fingerprints });
    let resp = request_with_retry(client, api_key, &url, Some(body)).await?;
    if !resp.status().is_success() {
        return Ok(HashMap::new());
    }
    let data: CfResponse<CfFingerprintData> = resp.json().await?;
    let mut result = HashMap::new();
    for m in data.data.exact_matches {
        let fingerprint = m.file.file_fingerprint as u32;
        if let Some(mv) = parse_cf_file(m.file, m.id, loader) {
            result.insert(fingerprint, (m.id.to_string(), mv));
        }
    }
    Ok(result)
}

/// Returns map from project_id -> project_name.
pub async fn get_mods_info(
    client: &Client,
    api_key: &str,
    mod_ids: &[u32],
) -> Result<HashMap<u32, String>> {
    if mod_ids.is_empty() {
        return Ok(HashMap::new());
    }
    let url = format!("{BASE_URL}/mods");
    let body = json!({ "modIds": mod_ids });
    let resp = request_with_retry(client, api_key, &url, Some(body)).await?;
    if !resp.status().is_success() {
        return Ok(HashMap::new());
    }
    let data: CfResponse<Vec<CfMod>> = resp.json().await?;
    Ok(data.data.into_iter().map(|m| (m.id, m.name)).collect())
}

/// Returns the latest compatible version for a project.
pub async fn get_latest_version(
    client: &Client,
    api_key: &str,
    mod_id: u32,
    mc_version: &str,
    loader: &str,
) -> Result<Option<ModVersion>> {
    let loader_id = loader_id(loader);
    let url = format!(
        "{BASE_URL}/mods/{mod_id}/files?gameVersion={mc_version}&modLoaderType={loader_id}&pageSize=1"
    );
    let resp = request_with_retry(client, api_key, &url, None).await?;
    if !resp.status().is_success() {
        return Ok(None);
    }
    #[derive(Deserialize)]
    struct CfFilesResponse {
        data: Vec<CfFile>,
    }
    let data: CfFilesResponse = resp.json().await.unwrap_or(CfFilesResponse { data: vec![] });
    Ok(data.data.into_iter().next().and_then(|f| parse_cf_file(f, mod_id, loader)))
}
